File: 3326-minimum-division-operations-to-make-array-non-decreasing/3326-minimum-division-operations-to-make-array-non-decreasing.py
"""
    Iterate through nums and find its proper divisors
    Find the greatest proper divisor

    Iterate through nums backwards, when we see element i-1 is greater than element i, divide element i-1 
    by its GPD. We use memoization 
    

     Runtime:
            O(n * log(m) * m/2), where m is the maximum number in nums.
"""

import math


class Solution:
    def minOperations(self, nums: List[int]) -> int:
        @cache
        def fingGpd(m):         
            for i in range(math.floor(m/2), 1, -1):
                if m % i == 0:
                    return i
            return 1

        counter = 0
        for i in range(len(nums) - 1, 0, -1):
            # How many times does this loop run?
            # In the worst case, gpd = 2 so that we make nums[i-1] smaller by the least amount.
            # In the worst case we repeat this log(nums[i-1])
            while nums[i] < nums[i - 1]:
                gpd = fingGpd(nums[i - 1])
                if gpd == 1:
                    return -1

                nums[i - 1] = nums[i - 1] / gpd
                counter += 1
        return counter

File: 3326-minimum-division-operations-to-make-array-non-decreasing/3326-minimum-division-operations-to-make-array-non-decreasing.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var minOperations = function(nums) {
    const n = nums.length;

    // Helper function to get all proper divisors of x
    const getProperDivisors = (x) => {
        const divisors = [];
        for (let i = 2; i <= Math.sqrt(x); i++) {
            if (x % i === 0) {
                divisors.push(i);
                if (i !== x / i) divisors.push(x / i);
            }
        }
        return divisors.sort((a, b) => b - a);  // Sort in descending order
    };

    let operations = 0;

    // Iterate from the second last element down to the first
    for (let i = n - 2; i >= 0; i--) {
        if (nums[i] <= nums[i + 1]) continue;

        const divisors = getProperDivisors(nums[i]);
        let divided = false;

        for (const divisor of divisors) {
            if (nums[i] / divisor <= nums[i + 1]) {
                nums[i] = Math.floor(nums[i] / divisor);
                operations++;
                divided = true;
                break;
            }
        }

        if (!divided) return -1;
    }

    return operations;
}

// Test case
//console.log(minOperations([288, 7]));  // Output: 1
